src/manifolds/SymplecticStiefel.jl

```diff
@@ -0,0 +1,554 @@
+    mul!(A4, p3', q2, -1, 1)    #A4  -= p3'q2
+    return A
+end
+
+@doc raw"""
+    Compute the exponential mapping from eq. 3.19 of Bendokat-Zimmermann.
+"""
+function exp!(M::SymplecticStiefel{n,k}, q, p, X) where {n,k}
+    Q = SymplecticMatrix(p, X)
+    pT_p = lu(p' * p) # ∈ ℝ^{2k × 2k}
+
+    C = pT_p \ X' # ∈ ℝ^{2k × 2n}
+
+    # Construct A-bar:
+    # First A-term: Q * (p^T * C^T) * Q
+    A_bar = rmul!(lmul!(Q, (p' * C')), Q)
+    A_bar .+= C * p
+
+    # Second A-term, use C-memory:
+    rmul!(C, Q') # C*Q^T -> C
+    C_QT = C
+
+    # Subtract C*Q^T*p*(pT_p)^{-1}*Q:
+    # A_bar is "star-skew symmetric" (A^+ = -A).
+    A_bar .-= rmul!((C_QT * p) / pT_p, Q)
+
+    # Construct H_bar:
+    # First H-term: Q * (C_QT * Q)' * Q -> Q * C' * Q = Q * (X / pT_p) * Q
+    H_bar = rmul!(lmul!(Q, rmul!(C_QT, Q)'), Q)
+    # Subtract second H-term:
+    H_bar .-= p * symplectic_inverse_times(M, p, H_bar)
+
+    # Construct Δ_bar in H_bar-memory:
+    H_bar .+= p * A_bar
+
+    # Rename H_bar -> Δ_bar.
+    Δ_bar = H_bar
+
+    γ_1 = Δ_bar - (1 / 2) .* p * symplectic_inverse_times(M, p, Δ_bar)
+    γ = [γ_1 -p] # ∈ ℝ^{2n × 4k}
+
+    Δ_bar_star = rmul!(Q' * Δ_bar', Q)
+    λ_1 = lmul!(Q', p * Q)
+    λ_2 = (Δ_bar_star .- (1 / 2) .* (Δ_bar_star * p) * λ_1')'
+    λ = [λ_1 λ_2] # ∈ ℝ^{2n × 4k}
+
+    Γ = [λ -γ] # ∈ ℝ^{2n × 8k}
+    Λ = [γ λ] # ∈ ℝ^{2n × 8k}
+
+    # At last compute the (8k × 8k) and (4k × 4k) matrix exponentials:
+    q .= Γ * (exp(Λ' * Γ)[:, (4k + 1):end]) * (exp(λ' * γ)[:, (2k + 1):end])
+    return q
+end
+
+@doc raw"""
+    retract!(::SymplecticStiefel, q, p, X, ::CayleyRetraction)
+
+Compute the Cayley retraction on the Symplectic Stiefel manifold, computed inplace
+of `q` from `p` along `X`.
+
+Given a point ``p \in \operatorname{SpSt}(2n, 2k)``, every tangent vector
+``X \in T_p\operatorname{SpSt}(2n, 2k)`` is of the form
+``X = \tilde{\Omega}p``, with
+````math
+    \tilde{\Omega} = \left(I_{2n} - \frac{1}{2}pp^+\right)Xp^+ -
+                     pX^+\left(I_{2n} - \frac{1}{2}pp^+\right) \in ℝ^{2n \times 2n},
+````
+as shown in Proposition 3.5 of [^Bendokat2021].
+Using this representation of ``X``, the Cayley retraction
+on ``\operatorname{SpSt}(2n, 2k)`` is defined pointwise as
+````math
+    \mathcal{R}_p(X) = \operatorname{cay}\left(\frac{1}{2}\tilde{\Omega}\right)p.
+````
+The operator ``\operatorname{cay}(A) = (I - A)^{-1}(I + A)`` is the Cayley transform.
+
+However, the computation of an ``2n \times 2n`` matrix inverse in the expression
+above can be reduced down to inverting a ``2k \times 2k`` matrix due to Proposition
+5.2 of [^Bendokat2021].
+
+Let ``A = p^+X`` and ``H = X - pA``. Then an equivalent expression for the Cayley
+retraction defined pointwise above is
+````math
+    \mathcal{R}_p(X) = -p + (H + 2p)(H^+H/4 - A/2 + I_{2k})^{-1}.
+````
+It is this expression we compute inplace of `q`.
+"""
+function retract!(M::SymplecticStiefel{n,k}, q, p, X, ::CayleyRetraction) where {n,k}
```

Comment:
We usually document the `retract` variant (which allocates memory and calls this version).

Reply:
Done.